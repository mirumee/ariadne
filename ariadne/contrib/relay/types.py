from collections import namedtuple
from typing import Any, Callable, Dict

from typing_extensions import TypeVar

from ariadne.contrib.relay.connection import RelayConnection

ConnectionResolver = TypeVar("ConnectionResolver", bound=Callable[..., RelayConnection])
GlobalIDTuple = namedtuple("GlobalIDTuple", ["type", "id"])
GlobalIDDecoder = Callable[[Dict[str, Any]], GlobalIDTuple]
