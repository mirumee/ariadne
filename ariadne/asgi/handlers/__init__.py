from .base import GraphQLHandler, GraphQLWebsocketHandler
from .http import GraphQLHTTPHandler
from .graphql_transport_ws import GraphQLTransportWSHandler
from .graphql_ws import GraphQLWSHandler


__all__ = [
    "GraphQLHandler",
    "GraphQLHTTPHandler",
    "GraphQLTransportWSHandler",
    "GraphQLWSHandler",
    "GraphQLWebsocketHandler",
]
