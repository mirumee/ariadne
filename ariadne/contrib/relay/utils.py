from base64 import b64decode, b64encode

from ariadne.contrib.relay.types import (
    GlobalIDTuple,
)


def decode_global_id(gid: str) -> GlobalIDTuple:
    return GlobalIDTuple(*b64decode(gid).decode().split(":"))


def encode_global_id(type_name: str, _id: str) -> str:
    return b64encode(f"{type_name}:{_id}".encode()).decode()
