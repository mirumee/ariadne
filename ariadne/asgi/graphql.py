from typing import Optional

from graphql import GraphQLSchema
from starlette.types import Receive, Scope, Send

from ..format_error import format_error
from ..types import (
    ContextValue,
    ErrorFormatter,
    RootValue,
    ValidationRules,
)
from .handlers import (
    GraphQLHTTPHandler,
    GraphQLWebsocketHandler,
    GraphQLWSHandler,
)


class GraphQL:
    def __init__(
        self,
        schema: GraphQLSchema,
        *,
        context_value: Optional[ContextValue] = None,
        root_value: Optional[RootValue] = None,
        validation_rules: Optional[ValidationRules] = None,
        debug: bool = False,
        introspection: bool = True,
        logger: Optional[str] = None,
        error_formatter: ErrorFormatter = format_error,
        http_handler: Optional[GraphQLHTTPHandler] = None,
        websocket_handler: Optional[GraphQLWebsocketHandler] = None,
    ):
        if http_handler:
            self.http_handler = http_handler
        else:
            self.http_handler = GraphQLHTTPHandler()

        if websocket_handler:
            self.websocket_handler = websocket_handler
        else:
            self.websocket_handler = GraphQLWSHandler()

        self.http_handler.configure(
            schema,
            context_value,
            root_value,
            validation_rules,
            debug,
            introspection,
            logger,
            error_formatter,
        )
        self.websocket_handler.configure(
            schema,
            context_value,
            root_value,
            validation_rules,
            debug,
            introspection,
            logger,
            error_formatter,
            http_handler=self.http_handler,
        )

    async def __call__(self, scope: Scope, receive: Receive, send: Send):
        if scope["type"] == "http":
            await self.http_handler.handle(scope=scope, receive=receive, send=send)
        elif scope["type"] == "websocket":
            await self.websocket_handler.handle(scope=scope, receive=receive, send=send)
        else:
            raise ValueError("Unknown scope type: %r" % (scope["type"],))
