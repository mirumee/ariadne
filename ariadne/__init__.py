from .enums import (
    EnumType,
    set_default_enum_values_on_schema,
    validate_schema_enum_values,
)
from .executable_schema import make_executable_schema
from .extensions import ExtensionManager
from .file_uploads import combine_multipart_data, upload_scalar
from .format_error import (
    format_error,
    get_error_extension,
    get_formatted_error_context,
    get_formatted_error_traceback,
)
from .graphql import graphql, graphql_sync, subscribe
from .inputs import InputType
from .interfaces import InterfaceType, type_implements_interface
from .load_schema import load_schema_from_path
from .objects import MutationType, ObjectType, QueryType
from .resolvers import (
    FallbackResolversSetter,
    SnakeCaseFallbackResolversSetter,
    fallback_resolvers,
    is_default_resolver,
    resolve_to,
    snake_case_fallback_resolvers,
)
from .scalars import ScalarType
from .schema_names import SchemaNameConverter, convert_schema_names
from .schema_visitor import SchemaDirectiveVisitor
from .subscriptions import SubscriptionType
from .types import Extension, SchemaBindable
from .unions import UnionType
from .utils import (
    convert_camel_case_to_snake,
    convert_kwargs_to_snake_case,
    gql,
    unwrap_graphql_error,
)

__all__ = [
    "EnumType",
    "Extension",
    "ExtensionManager",
    "FallbackResolversSetter",
    "InputType",
    "InterfaceType",
    "MutationType",
    "ObjectType",
    "QueryType",
    "ScalarType",
    "SchemaBindable",
    "SchemaDirectiveVisitor",
    "SchemaNameConverter",
    "SnakeCaseFallbackResolversSetter",
    "SubscriptionType",
    "UnionType",
    "combine_multipart_data",
    "convert_camel_case_to_snake",
    "convert_kwargs_to_snake_case",
    "convert_schema_names",
    "fallback_resolvers",
    "format_error",
    "get_error_extension",
    "get_formatted_error_context",
    "get_formatted_error_traceback",
    "gql",
    "graphql",
    "graphql_sync",
    "is_default_resolver",
    "load_schema_from_path",
    "make_executable_schema",
    "resolve_to",
    "set_default_enum_values_on_schema",
    "snake_case_fallback_resolvers",
    "subscribe",
    "type_implements_interface",
    "unwrap_graphql_error",
    "upload_scalar",
    "validate_schema_enum_values",
]
