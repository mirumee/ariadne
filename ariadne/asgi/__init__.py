from ..types import (
    Extensions,
    MiddlewareList,
    Middlewares,
    OnComplete,
    OnConnect,
    OnDisconnect,
    OnOperation,
    Operation,
    WebSocketConnectionError,
)
from .graphql import GraphQL

__all__ = [
    "Extensions",
    "GraphQL",
    "MiddlewareList",
    "Middlewares",
    "OnComplete",
    "OnConnect",
    "OnDisconnect",
    "OnOperation",
    "Operation",
    "WebSocketConnectionError",
]
