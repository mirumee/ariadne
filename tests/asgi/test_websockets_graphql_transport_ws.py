# pylint: disable=not-context-manager
import pytest
from graphql import parse
from graphql.language import OperationType
from starlette.testclient import TestClient
from starlette.websockets import WebSocketDisconnect

from ariadne.asgi import GraphQL
from ariadne.asgi.handlers import GraphQLTransportWSHandler
from ariadne.types import WebSocketConnectionError
from ariadne.utils import get_operation_type


def test_field_can_be_subscribed_using_websocket_connection_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE
        assert response["id"] == "test1"


def test_field_can_be_subscribed_using_unnamed_operation_in_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": None,
                    "query": "subscription { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE
        assert response["id"] == "test1"


def test_field_can_be_subscribed_using_named_operation_in_websocket_connection_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "PingTest",
                    "query": "subscription PingTest { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE
        assert response["id"] == "test1"


def test_query_can_be_executed_using_websocket_connection_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test2",
                "payload": {
                    "operationName": None,
                    "query": "query Hello($name: String){ hello(name: $name) }",
                    "variables": {"name": "John"},
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test2"
        assert response["payload"]["data"] == {"hello": "Hello, John!"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE
        assert response["id"] == "test2"


def test_immediate_disconnect_on_invalid_type_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json({"type": "INVALID_TYPE"})
        with pytest.raises(WebSocketDisconnect):
            assert ws.receive_json()


def test_complete_graphql_transport_ws(client_graphql_transport_ws):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_COMPLETE, "id": "test1"})
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"


def test_pong_graphql_transport_ws(client_graphql_transport_ws):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_PONG,
            }
        )


def test_custom_websocket_on_connect_is_called_graphql_transport_ws(schema):
    test_payload = None

    def on_connect(websocket, payload):
        assert payload == test_payload
        websocket.scope["payload"] = payload

    websocket_handler = GraphQLTransportWSHandler(on_connect=on_connect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        assert ws.scope["payload"] == test_payload


def test_custom_websocket_on_connect_is_called_with_payload_graph_graphql_transport_ws(
    schema,
):
    test_payload = {"test": "ok"}

    def on_connect(websocket, payload):
        assert payload == test_payload
        websocket.scope["payload"] = payload

    websocket_handler = GraphQLTransportWSHandler(on_connect=on_connect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT,
                "payload": test_payload,
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        assert ws.scope["payload"] == test_payload


def test_custom_websocket_on_connect_is_awaited_if_its_async_graphql_transport_ws(
    schema,
):
    test_payload = {"test": "ok"}

    async def on_connect(websocket, payload):
        assert payload == test_payload
        websocket.scope["payload"] = payload

    websocket_handler = GraphQLTransportWSHandler(on_connect=on_connect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT,
                "payload": test_payload,
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        assert ws.scope["payload"] == test_payload


def test_error_in_custom_websocket_on_connect_closes_connection_graphql_transport_ws(
    schema,
):
    def on_connect(websocket, payload):
        raise ValueError("Oh No!")

    websocket_handler = GraphQLTransportWSHandler(on_connect=on_connect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        with pytest.raises(WebSocketDisconnect):
            ws.receive_json()


def test_custom_websocket_connection_error_closes_connection_graphql_transport_ws(
    schema,
):
    def on_connect(websocket, payload):
        raise WebSocketConnectionError({"msg": "Token required", "code": "auth_error"})

    websocket_handler = GraphQLTransportWSHandler(on_connect=on_connect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        with pytest.raises(WebSocketDisconnect):
            ws.receive_json()


def test_custom_websocket_on_operation_is_called_graphql_transport_ws(schema):
    def on_operation(websocket, operation):
        assert operation.name == "TestOp"
        websocket.scope["on_operation"] = True

    websocket_handler = GraphQLTransportWSHandler(on_operation=on_operation)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}

        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE
        assert ws.scope["on_operation"] is True


def test_custom_websocket_on_operation_is_awaited_if_its_async_graphql_transport_ws(
    schema,
):
    async def on_operation(websocket, operation):
        assert operation.name == "TestOp"
        websocket.scope["on_operation"] = True

    websocket_handler = GraphQLTransportWSHandler(on_operation=on_operation)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_COMPLETE, "id": "test1"})
        assert ws.scope["on_operation"] is True


def test_error_in_custom_websocket_on_operation_is_handled_graphql_transport_ws(schema):
    async def on_operation(websocket, operation):
        raise ValueError("Oh No!")

    websocket_handler = GraphQLTransportWSHandler(on_operation=on_operation)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}


def test_custom_websocket_on_complete_is_called_on_disconnect_graphql_transport_ws(
    schema,
):
    def on_complete(websocket, operation):
        assert operation.name == "TestOp"
        websocket.scope["on_complete"] = True

    websocket_handler = GraphQLTransportWSHandler(on_complete=on_complete)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )

    assert ws.scope["on_complete"] is True


def test_custom_websocket_on_complete_is_called_on_operation_complete_grapqhl_transport_ws(
    schema,
):
    def on_complete(websocket, operation):
        assert operation.name == "TestOp"
        websocket.scope["on_complete"] = True

    websocket_handler = GraphQLTransportWSHandler(on_complete=on_complete)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}

        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE

        assert ws.scope["on_complete"] is True


def test_custom_websocket_on_complete_is_awaited_if_its_async_graphql_transport_ws(
    schema,
):
    async def on_complete(websocket, operation):
        assert operation.name == "TestOp"
        websocket.scope["on_complete"] = True

    websocket_handler = GraphQLTransportWSHandler(on_complete=on_complete)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE

    assert ws.scope["on_complete"] is True


def test_error_in_custom_websocket_on_complete_is_handled_graphql_transport_ws(schema):
    async def on_complete(websocket, operation):
        raise ValueError("Oh No!")

    websocket_handler = GraphQLTransportWSHandler(on_complete=on_complete)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {
                    "operationName": "TestOp",
                    "query": "subscription TestOp { ping }",
                },
            }
        )
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_NEXT
        assert response["id"] == "test1"
        assert response["payload"]["data"] == {"ping": "pong"}
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_COMPLETE


def test_custom_websocket_on_disconnect_is_called_on_invalid_operation_graphql_transport_ws(
    schema,
):
    def on_disconnect(websocket):
        websocket.scope["on_disconnect"] = True

    websocket_handler = GraphQLTransportWSHandler(on_disconnect=on_disconnect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json({"type": "INVALID"})
        assert "on_disconnect" not in ws.scope

    assert ws.scope["on_disconnect"] is True


def test_custom_websocket_on_disconnect_is_called_on_connection_close_graphql_transport_ws(
    schema,
):
    def on_disconnect(websocket):
        websocket.scope["on_disconnect"] = True

    websocket_handler = GraphQLTransportWSHandler(on_disconnect=on_disconnect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        assert "on_disconnect" not in ws.scope

    assert ws.scope["on_disconnect"] is True


def test_custom_websocket_on_disconnect_is_awaited_if_its_async_graphql_transport_ws(
    schema,
):
    async def on_disconnect(websocket):
        websocket.scope["on_disconnect"] = True

    websocket_handler = GraphQLTransportWSHandler(on_disconnect=on_disconnect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json({"type": "INVALID"})
        assert "on_disconnect" not in ws.scope

    assert ws.scope["on_disconnect"] is True


def test_error_in_custom_websocket_on_disconnect_is_handled_graphql_transport_ws(
    schema,
):
    async def on_disconnect(websocket):
        raise ValueError("Oh No!")

    websocket_handler = GraphQLTransportWSHandler(on_disconnect=on_disconnect)
    app = GraphQL(schema, websocket_handler=websocket_handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json({"type": "INVALID"})


def test_too_many_connection_init_messages_graphql_transport_ws(
    schema,
):
    handler = GraphQLTransportWSHandler()
    app = GraphQL(schema, websocket_handler=handler)
    client = TestClient(app)

    with client.websocket_connect("/", ["graphql-transport-ws"]) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        response = ws.receive_json()
        assert response["type"] == GraphQLTransportWSHandler.GQL_CONNECTION_ACK
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        with pytest.raises(WebSocketDisconnect, match="4429"):
            ws.receive_json()


def test_get_operation_type():
    graphql_document = parse("subscription ping {ping} query other { dummy }")
    operation_type = get_operation_type(graphql_document, "ping")
    assert operation_type == OperationType.SUBSCRIPTION

    operation_type = get_operation_type(graphql_document, "other")
    assert operation_type == OperationType.QUERY

    operation_type = get_operation_type(graphql_document)
    assert operation_type == OperationType.SUBSCRIPTION


def test_connection_not_acknowledged_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        with pytest.raises(WebSocketDisconnect, match="4401"):
            ws.receive_json()


def test_duplicate_operation_id_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        ws.send_json(
            {
                "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                "id": "test1",
                "payload": {"query": "subscription { ping }"},
            }
        )
        ws.receive_json()
        with pytest.raises(WebSocketDisconnect, match="4409"):
            ws.receive_json()


def test_invalid_operation_id_is_handled_graphql_transport_ws(
    client_graphql_transport_ws,
):
    with client_graphql_transport_ws.websocket_connect(
        "/", ["graphql-transport-ws"]
    ) as ws:
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
        ws.send_json({"type": GraphQLTransportWSHandler.GQL_COMPLETE, "id": "test1"})
        ws.receive_json()


def test_schema_not_set_graphql_transport_ws(
    client_graphql_transport_ws,
):

    client_graphql_transport_ws.app.websocket_handler.schema = None
    with pytest.raises(TypeError):
        with client_graphql_transport_ws.websocket_connect(
            "/", ["graphql-transport-ws"]
        ) as ws:
            ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
            ws.send_json(
                {
                    "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                    "id": "test1",
                    "payload": {"query": "subscription { ping }"},
                }
            )


def test_http_handler_not_set_graphql_transport_ws(
    client_graphql_transport_ws,
):

    client_graphql_transport_ws.app.websocket_handler.http_handler = None
    with pytest.raises(TypeError):
        with client_graphql_transport_ws.websocket_connect(
            "/", ["graphql-transport-ws"]
        ) as ws:
            ws.send_json({"type": GraphQLTransportWSHandler.GQL_CONNECTION_INIT})
            ws.send_json(
                {
                    "type": GraphQLTransportWSHandler.GQL_SUBSCRIBE,
                    "id": "test2",
                    "payload": {
                        "operationName": None,
                        "query": "query Hello($name: String){ hello(name: $name) }",
                        "variables": {"name": "John"},
                    },
                }
            )
