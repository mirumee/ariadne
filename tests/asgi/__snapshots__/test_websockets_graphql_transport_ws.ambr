# serializer version: 1
# name: test_invalid_query_error_is_handled_using_websocket_connection_graphql_transport_ws
  list([
    dict({
      'locations': list([
        dict({
          'column': 17,
          'line': 1,
        }),
      ]),
      'message': "Cannot query field 'error' on type 'Query'.",
    }),
    dict({
      'locations': list([
        dict({
          'column': 23,
          'line': 1,
        }),
      ]),
      'message': "Cannot query field 'other' on type 'Query'.",
    }),
  ])
# ---
